plugins/flatpak/gs-flatpak.c

```diff
@@ -2477,18 +2477,82 @@ gs_flatpak_app_remove_source (GsFlatpak *self,
 	return TRUE;
 }
 
+/*
+ * gs_flatpak_get_services_app_for_runtime:
+ * Specific to Endless runtimes, gets a GsApp reference for the related
+ * "EknServices" app that provides system services for apps that use the
+ * given runtime.
+ *
+ * Returns NULL if no services app was required or there was an error.
+ * Does not take a GError since any error should be recoverable. However, in
+ * order to make sure the operation was not cancelled, you should check the
+ * status of the GCancellable after calling this.
+ */
+static GsApp *
+gs_flatpak_get_services_app_for_runtime (GsFlatpak *self, GsApp *runtime,
+                                         GCancellable *cancellable)
+{
+	const gchar *runtime_id, *runtime_branch, *services_branch, *arch;
+	g_autofree gchar *services_id = NULL;
+	g_autofree gchar *description = NULL;
+	g_autoptr(FlatpakRef) services_ref = NULL;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GsApp) retval = NULL;
+
+	runtime_id = gs_app_get_id (runtime);
+	runtime_branch = gs_app_get_branch (runtime);
+
+	if ((g_strcmp0 (runtime_id, "com.endlessm.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "eos3.1") == 0) ||
+	    (g_strcmp0 (runtime_id, "com.endlessm.apps.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "1") == 0)) {
+		services_id = g_strdup ("com.endlessm.EknServices");
+		services_branch = "eos3";
+	} else if (g_strcmp0 (runtime_id, "com.endlessm.apps.Platform") == 0) {
+		if (g_strcmp0 (runtime_branch, "master") == 0)
+			services_id = g_strdup ("com.endlessm.EknServices2");
+		else
+			services_id = g_strdup_printf ("com.endlessm.EknServices%s",
+			                               runtime_branch);
+		services_branch = "stable";
+	} else {
+		/* Runtime doesn't require an EknServices app */
+		return NULL;
+	}
+
+	/* Construct a GsApp for the EknServices we determined we needed */
+	arch = gs_flatpak_app_get_ref_arch (runtime);
+	description = g_strdup_printf ("app/%s/%s/%s", services_id, arch, services_branch);
+	services_ref = flatpak_ref_parse (description, &error);
+	if (!services_ref) {
+		g_critical ("Invalid flatpak ref description %s: %s", description,
+		            error->message);
+		return NULL;
+	}
+
+	retval = gs_flatpak_create_app (self, services_ref);
+	gs_app_set_origin (retval, gs_app_get_origin (runtime));
```

Comment:
If you do this then you'll make it look for the services app in the runtime's remote, and this is why you're getting the failure you've reported in [here](https://phabricator.endlessm.com/T19806#504381). Also, by setting the origin (and if the origin is not _NULL_), the next line will do nothing as it first verifies if an origin is already set before it proceeds to look for it.
So maybe all that's needed is to delete this line.

Reply:
OMG, I stared at this for so long and didn't notice that I had failed to remove this line! Thanks!